File: ReadingTool.Services/LatexParserService.cs
#region License
// LatexParserService.cs is part of ReadingTool.Services
// 
// ReadingTool.Services is free software: you can redistribute it and/or modify
// it under the terms of the GNU Affero General Public License as published by
// the Free Software Foundation, either version 3 of the License, or
// (at your option) any later version.
// 
// ReadingTool.Services is distributed in the hope that it will be useful,
// but WITHOUT ANY WARRANTY; without even the implied warranty of
// MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
// GNU Affero General Public License for more details.
// 
// You should have received a copy of the GNU Affero General Public License
// along with ReadingTool.Services. If not, see <http://www.gnu.org/licenses/>.
// 
// Copyright (C) 2013 Travis Watt
#endregion

using System;
using System.IO;
using System.Linq;
using System.Text;
using System.Xml;
using System.Xml.Linq;
using System.Xml.Xsl;
using ReadingTool.Entities;

namespace ReadingTool.Services
{
    public class LatexParserService : DefaultParserService
    {
        public LatexParserService()
            : base()
        {
        }

        protected virtual Tuple<int, XDocument> ClassTerms(XDocument document)
        {
            var totalTerms = document.Descendants("t").Count(x => x.Attribute("type").Value == "term");

            //Multilength
            /*
            foreach(var multi in _terms)
            {
                string lower = multi.TermPhrase.ToLowerInvariant();

                var first = lower.Split(' ').First();
                var partials = document
                    .Descendants("t")
                    .Where(x =>
                        x.Attribute("type").Value == "term" &&
                        x.Attribute("lower").Value.Equals(first, StringComparison.InvariantCultureIgnoreCase));

                if(partials == null || !partials.Any()) continue;

                foreach(var element in partials)
                {
                    var nodeString = element.Attribute("lower").Value + " " +
                        string.Join(
                            " ",
                            (element.ElementsAfterSelf().Where(x => x.Attribute("type").Value == "term").Take(multi.Length - 1))
                            .Select(x => x.Attribute("lower").Value)
                            );

                    if(!lower.Equals(nodeString, StringComparison.InvariantCultureIgnoreCase))
                    {
                        continue;
                    }

                    var node = new XElement("multi");
                    node.Value = multi.TermPhrase;

                    if(multi.State == TermState.Unknown)
                    {
                        node.SetAttributeValue("box", "box" + multi.Box);
                    }

                    node.SetAttributeValue("length", multi.Length);
                    node.SetAttributeValue("lower", lower);
                    node.SetAttributeValue("state", Constants.TermStatesToClass[multi.State]);
                    node.SetAttributeValue("id", multi.Id.ToString());
                    node.SetAttributeValue("data", multi.Definition);

                    element.Add(node);
                }
            }
            */
            var elements = document.Descendants("t").Where(x => x.Attribute("type").Value == "term");
            var termsAsDict = _singleTerms.ToDictionary(x => x.Phrase.ToLowerInvariant(), x => new { State = x.State, FullDefinition = x.FullDefinition });

            foreach(var element in elements)
            {
                var lower = element.Attribute("lower").Value;

                if(termsAsDict.ContainsKey(lower))
                {
                    element.SetAttributeValue("state", Term.TermStateToClass(termsAsDict[lower].State));
                    element.SetAttributeValue("data", termsAsDict[lower].FullDefinition);
                }
            }

            return new Tuple<int, XDocument>(totalTerms, document);
        }

        protected override XDocument CreateStats(int totalWords, XDocument document)
        {
            return document;
        }

        protected override string ApplyTransform(XDocument document)
        {
            string xslText = XsltResource.Read_Latex;

            XmlWriterSettings settings = new XmlWriterSettings();
            settings.OmitXmlDeclaration = true;
            settings.ConformanceLevel = ConformanceLevel.Fragment;

            StringBuilder sb = new StringBuilder();
            using(StringWriter sw = new StringWriter(sb))
            {
                using(XmlWriter writer = XmlWriter.Create(sw, settings))
                {
                    XslCompiledTransform xslt = new XslCompiledTransform();
                    xslt.Load(XmlReader.Create(new StringReader(xslText)));
                    xslt.Transform(document.CreateReader(), writer);
                }
            }

            return sb.ToString();
        }
    }
}

File: app/classes/services/LatexParserService.php
<?php

namespace RT\Services;

use App\Models\Language;
use App\Models\Text;
use App\Models\Term;
use \User;
use RT\Core\Splitter;

class LatextParserService extends ParserService {
    protected function classTerms($xml) {
        $termsAsDict = array();
        
        foreach($this->terms as $t) {
            if($t->length!=1 || $t->state!='unknown') {
                continue;
            }
            
            $termsAsDict[$t->phraseLower] = 
                    array(
                        'state'=>$t->state,
                        'definition'=>$t->fullDefinition()
                    );
        }
        
        $xpath = new \DomXpath($xml);
        $query = $xpath->query("//t[@type='term']");
        
        foreach ($query as $node) {
            $lower = $node->getAttribute("lower");
            
            if(array_key_exists($lower, $termsAsDict)) {
                $term = $termsAsDict[$lower];
                
                $node->setAttribute("state", $this->termToStateClass($term['state']));
                $node->setAttribute("data", $term['definition']);
            }
        }
        
        return $query->length;
    }

    protected function createStats($totalTerms, $document) {
        return $document;
    }
    
    protected function applyTransform($xml) {
        $xsl = new \DOMDocument;
        
        $contents = $xsl->load(public_path() . '/xslt/latex.xsl');
        $proc = new \XSLTProcessor;
        $proc->importStyleSheet($xsl);
        
        return $proc->transformToXML($xml);
    }
}
